### PR TITLE
C# 9 reference update: lambda discard parameters

### PR DESCRIPTION
**Pull request by pkulikov**
Contributes to #20577

### CODE REVIEW

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 1]
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
...
+(input-parameters) => { <sequence-of-statements> }
+```
+
+The body of a statement lambda can consist of any number of statements; however, in practice there are typically no more than two or three.
+
+[!code-csharp-interactive[statement lambda](snippets/lambda-expressions/ExpressionAndStatementLambdas.cs#StatementLambda)]
+
+You cannot use statement lambdas to create expression trees.
+
+## Input parameters of a lambda expression
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/lambda-expressions.md:69]: That information was inside the section about expression lambdas, while it's true for all lambdas. So, I've extracted the section.

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 2]
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
   
 ## Expression lambdas
 
-A lambda expression with an expression on the right side of the `=>` operator is called an *expression lambda*. Expression lambdas are used extensively in the construction of [expression trees](../../programming-guide/concepts/expression-trees/index.md). An expression lambda returns the result of the expression and takes the following basic form:
+A lambda expression with an expression on the right side of the `=>` operator is called an *expression lambda*. An expression lambda returns the result of the expression and takes the following basic form:
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/lambda-expressions.md:47]: Removed that sentence as it breaks the flow of the paragraph; moreover, that information is mentioned earlier in the article.

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 3]
```diff
@@ -216,10 +229,8 @@ The following rules apply to variable scope in lambda expressions:
 
 For more information, see the [Anonymous function expressions](~/_csharplang/spec/expressions.md#anonymous-function-expressions) section of the [C# language specification](~/_csharplang/spec/introduction.md).
 
-## Featured book chapter
+For more information about lambda discard parameters, see the [feature proposal note](~/_csharplang/proposals/csharp-9.0/lambda-discard-parameters.md)
 
-[Delegates, Events, and Lambda Expressions](/previous-versions/visualstudio/visual-studio-2008/ff518994(v=orm.10)) in [C# 3.0 Cookbook, Third Edition: More than 250 solutions for C# 3.0 programmers](/previous-versions/visualstudio/visual-studio-2008/ff518995(v=orm.10))
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/lambda-expressions.md:221]: This article has been updated considerably during the last years; I guess, it's very far from that original chapter now (which is for C# 3)

**Review by pkulikov — COMMENTED** [review]
@BillWagner I've left a number of clarifying comments for some of the changes

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 4]
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
...
 
 [!code-csharp[explicitly typed parameters](snippets/lambda-expressions/ExpressionAndStatementLambdas.cs#ExplicitlyTypedParameters)]
 
 Input parameter types must be all explicit or all implicit; otherwise, a [CS0748](../../misc/cs0748.md) compiler error occurs.
 
-The body of an expression lambda can consist of a method call. However, if you are creating expression trees that are evaluated outside the context of the .NET common language runtime, such as in SQL Server, you should not use method calls in lambda expressions. The methods will have no meaning outside the context of the .NET common language runtime.
+Beginning with C# 9.0, you can use [discards](../../discards.md) to specify two or more input parameters of a lambda expression that aren't used in the expression:
 
-## Statement lambdas
+[!code-csharp[discard parameters](snippets/lambda-expressions/ExpressionAndStatementLambdas.cs#Discards)]
```
Comment (IEvangelist) [inline docs/csharp/language-reference/operators/lambda-expressions.md:91]:
```suggestion
:::code language="csharp" source="snippets/lambda-expressions/ExpressionAndStatementLambdas.cs" id="SnippetDiscards":::
```

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 5]
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
...
 
-A statement lambda resembles an expression lambda except that the statement(s) is enclosed in braces:
+That feature may be useful when you use a lambda expression to [provide an event handler](../../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
```
Comment (IEvangelist) [inline docs/csharp/language-reference/operators/lambda-expressions.md:93]:
```suggestion
Lambda discard parameters may be useful when you use a lambda expression to [provide an event handler](../../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
```

**Review by IEvangelist — COMMENTED** [review]
These are awesome changes, well done. I had a few bits of feedback and I'll let @BillWagner handle it from here. As always, thank you @pkulikov - great job!

File: docs/csharp/language-reference/operators/lambda-expressions.md [hunk 6]
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
...
 (input-parameters) => expression
 ```
 
-The parentheses are optional only if the lambda has one input parameter; otherwise they are required.
+The body of an expression lambda can consist of a method call. However, if you are creating [expression trees](../../programming-guide/concepts/expression-trees/index.md) that are evaluated outside the context of the .NET common language runtime, such as in SQL Server, you should not use method calls in lambda expressions. The methods will have no meaning outside the context of the .NET common language runtime.
 
-Specify zero input parameters with empty parentheses:  
+## Statement lambdas
+
+A statement lambda resembles an expression lambda except that the statement(s) is enclosed in braces:
```
[the remark this answers is not in the document]
Comment (pkulikov) [reply docs/csharp/language-reference/operators/lambda-expressions.md:57]: @IEvangelist  Should it be "its statements" instead of "their"?

**Comment by pkulikov** [comment]
@IEvangelist Thank you for review and suggestions.

File: docs/csharp/language-reference/operators/lambda-expressions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/csharp/language-reference/operators/lambda-expressions.md:57]: I'd be good with that, although I believe that using their in this context is fine too.

File: docs/csharp/language-reference/operators/lambda-expressions.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/lambda-expressions.md:221]: Yes, that's true.

Editorial tidbit: When I last updated Effective C# and More Effective C# (for versions 6 & 7, respectively), one of my reviewers made the point of commenting whenever I missed a reference to a "new" feature from C# 3 (the previous updates.)

**Review by BillWagner — COMMENTED** [review]
This is great @pkulikov 


I have only one general comment beyond what @IEvangelist mentioned (which have already been addressed).

I'd like to see a link to these changes from the C# 9 what's new article. That helps readers find more detailed information.

**Comment by pkulikov** [comment]
@BillWagner thank you. I've updated the What's new article with the links (also added the links that should have been added in the previous similar PRs).

**Review by BillWagner — APPROVED** [review]
You are awesome!   🎆 Thank you.

I'll approve this and :shipit: now.

**Pull request merged**